fix: notify users if reboot is required

lucasmoura opened the pull request:
## Proposed Commit Message
fix: notify users if reboot is required

When upgrading packages during ua fix we may install a package that requires a system reboot to fully fix the issue at hand. We are now alerting the user if a reboot is required when running ua fix.

Fixes: #1476

PS: The approach used here some shortcomings:

* If the user has installed a package that requires reboot and then run `ua fix`, we will not known if the reboot is need because of ua fix or the package installed by the user.
* If we detect that a reboot is required, should we say that the CVE/USN was not fixed ? I am using that approach, but I am not sure if that is the intended behavior.

## Test Steps
Run the new unittest added

## Desired commit type::
<!-- put an `x` in one merge type to allow the reviewer to merge for you -->
 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

blackboxsw commented on uaclient/security.py:
```diff
@@ -859,6 +859,17 @@ def prompt_for_affected_packages(
     if unfixed_pkgs:
         print(_format_unfixed_packages_msg(unfixed_pkgs))
 
+    if util.should_reboot():
```
@lucasmoura, we are adding a second call-site which does util.should_reboot and messaging when reboot is needed. What do you think about refactoring util.should_reboot to accept an optional message: str param which we can pass in instead of having a [Entitlement separate class method.check_for_reboot_msg](https://github.com/canonical/ubuntu-advantage-client/blob/master/uaclient/entitlements/repo.py#L85)

blackboxsw replied:
Also, within this conditional if true, I think we also want to call to `cfg.add_notice("", status.MESSAGE_ENABLE_REBOOT_REQUIRED_TMPL.format(operation="fix operation"))`


If we do that we also need a cfg.remove_notice in reboot_cmds.py that'll clean up the reboot required message.
When you peek at the remove_notices in reboot_cmds.py let us know if you think we need to make better use descr_regex to remove any matching reboot message from notices

lucasmoura replied:
@blackboxsw I agree on the notices approach, since the user may not reboot immediately after running the `ua fix` operation. However, I don't think allowing `should_reboot` to print messages will make the code better.

We would change the current code from:
```python
if util.should_reboot():
    print(
        status.MESSAGE_ENABLE_REBOOT_REQUIRED_TMPL.format(
            operation="fix operation"
        )    
    )    
    print(
        status.MESSAGE_SECURITY_ISSUE_NOT_RESOLVED.format(issue=issue_id)
    )    
    return
```

to:
```python
msg =  status.MESSAGE_ENABLE_REBOOT_REQUIRED_TMPL.format(
            operation="fix operation"
        )
        + "\n"
       + status.MESSAGE_SECURITY_ISSUE_NOT_RESOLVED.format(issue=issue_id)
if util.should_reboot(msg):
    return
```

My point is that I think it is better to make `should_reboot` be as simple as possible and we drive want we want to do outside of it. Even though we have that same behavior on FIPS, I still think it is better to have the output separate from the `should_reboot` function

blackboxsw replied:
+1 thanks for your suggestion/response here. Agreed.

blackboxsw commented:
Confirmed current CI failure on this is           raise EOFError() on lxc init. known intermittent issue which hit bionic run this time. Doesn't affect local integration run.

blackboxsw reviewed (changes requested):
Thanks Lucas. Generally looks pretty good, I think we should also provide notices related to this too so that `ua status` reports about this reboot beyond the transative message printed to console.  Maybe it's worth us thinking a bit about handling notice cleanup a bit more generically in  reboot_cmds to cleanse any messages of type "reboot required".  I'm not really sure on that second part, but food for thought.

lucasmoura commented:
@blackboxsw I have changed the approach a little bit. Instead of clearing the notices on the `reboot_cmd` script and I am now configuring `ua status` to check if a reboot operation is needed. If it is not, we clear all reboot notices.

I think this is a better solution since it is easier to test than pushing this cleanup to the `reboot_cmd` script. However, it will make the status operation slower. But since we don't have that many notices at the moment, this new cleanup step should not be that much of a burden in the `status` operation

blackboxsw commented on uaclient/config.py:
```diff
@@ -526,6 +526,18 @@ def status(self, show_beta=False) -> "Dict[str, Any]":
+        # Try to remove all reboot notices if not applicable
+        if not util.should_reboot():
+            self.remove_notice("", status.MESSAGE_FIPS_REBOOT_REQUIRED)
+            self.remove_notice("", status.MESSAGE_FIPS_DISABLE_REBOOT_REQUIRED)
```
I think we are missing status.MESSAGE_LIVEPATCH_LTS_REBOOT_REQUIRED too and I think this also means we can drop the remove_notices logic from reboot_cmds.py too

lucasmoura replied:
For the `FIPS entitlements` case this will not happen because we check in `ua status`  the existence of the `/var/run/reboot-required` which indeed exists as can be seen here:
https://github.com/canonical/ubuntu-advantage-client/blob/master/uaclient/entitlements/fips.py#L56

Now, for the FIPS PRO case, yeah, this will happen. But we can avoid that by also checking if the marker file is not present.
In that case, we will guarantee that the FIPS PRO case will be handled by the reboot script only and the rest of the cases will be handled by status.

What do you think ?

lucasmoura commented:
@blackboxsw I have updated the code to only remove the `fix` notice during `ua status`. The reason for that is because we are already managing the other notices on separate places. For the FIPS ones that I have added, I have found that we are already cleaning them up here:

https://github.com/canonical/ubuntu-advantage-client/blob/master/uaclient/entitlements/fips.py#L187
